### ISSUE TITLE
Deploying as a GitHub Project Page based site needs canonifyurls = true

### ISSUE THREAD

**Issue by abulka** [gohugoio/hugo#7714]
Deploying as a GitHub `Project Page` based site needs `canonifyurls = true` because all static image references generated from markdown files are missing the sub-path `projectname/` component of the `baseURL`, which means those images break on deployment and cannot be found.

Seeing that deploying to a GitHub `Project Page` based site would be a fairly common thing to want to do, and assuming that many people are using markdown to author their hugo content, it would be good if image references worked properly on such a deployment. I believe this also affects GitLab Project Pages (the one with `username.gitlab.io/projectname`). 

This issue is also related to an issue that affects some themes, and unless the theme writer has cleverly adapted their theme, some themes break on deployment to GitHub `Project Page` based sites - but that is another issue - here I am focussing on normal users expecting the markdown image references to work upon deployment.

<pre>
$ hugo version
Hugo Static Site Generator v0.75.1/extended darwin/amd64 BuildDate: unknown
</pre>

### Does this issue reproduce with the latest release?

yes

There seem to be a whole bunch of issues and posts relating to this issue in some way.

- https://github.com/gohugoio/hugo/issues/5736
- https://github.com/gohugoio/hugo/issues/5226
- https://github.com/hossainemruz/toha/issues/89
- https://discourse.gohugo.io/t/baseurl-path-subdir-sometimes-ignored-by-relurl-absurl/11512/11
- https://azizunsal.github.io/blog/post/hugo-problem/
- https://github.com/gohugoio/hugo/issues/2194
- https://discourse.gohugo.io/t/image-path/1721/14
- https://stackoverflow.com/questions/55343573/generated-links-broken-in-hugo-blogdown-when-publishing-to-domain-with-subfolder

## Repro

E.g. The first image url refers to an image inside `content/blog/images`, and the second image url refers to a static image in `/static/images`. 
```markdown
![png your image](/blog/images/fred.png)
![png your image](/images/fred-uml.png)
```

Create images at `/blog/images/fred.png` and `static/images/fred-uml.png`.

Run `hugo`

Assuming `baseURL = "https://user.github.io/projectname/"` and `canonifyurls = false` (the hugo default) then after static site generation I get paths which are missing the essential `projectname/` path:
```html
<img src="blog/images/fred.png" alt="png your image">
<img src="images/fred-uml.png" alt="png your image">
```

Leaving the leading `/` off the image references doesn't help, and in fact breaks the local serving of those images.

## Solutions

Setting `canonifyurls = true` in config fixes the problem, but means all my urls become long, absolute urls - arguably a bit verbose and inelegant. Plus its not the default for hugo, so a lot of people are going to hit this trap and have to figure it out.

If you want to leave `canonifyurls = false` then another solution I came up with is to add a custom shortcode file to your site:  `layouts/shortcodes/img.html` containing:
```html
{{ .Get 0 | absURL }}
```
The markdown image reference syntax then needs to be:
```markdown
![png your image]({{< img "images/test-uml.png" >}})
```

**Comment by abulka** [comment]
I just found out that this issue also affects references to **style sheets** and probably other static content (not just images).  For example assuming `baseURL = "https://abulka.github.io/hugo-bare1/"` and my `themes/diagnostic-andy/layouts/_default/baseof.html` contains:
```html
<link rel="stylesheet" type="text/css" href="/css/style.css">
```
after a `hugo` command to generate the site, I get the incorrect 
```html
<link rel="stylesheet" type="text/css" href="/css/style.css">      <-- missing  /hugo-bare1/
```
which is wrong when deployed to a github **project pages** site. Hugo should arguably also generate the missing `/hugo-bare1` path like this:
```html
<link rel="stylesheet" type="text/css" href="/hugo-bare1/css/style.css">
```

## What this means
Thus the `img` shortcode workaround seems like the beginning of a rabbit hole of more shortcode workarounds needed... Looks like the simplest solution is just to bite the bullet and set `canonifyurls = true` in config. Which then generates the verbose, but correct:
```html
<link rel="stylesheet" type="text/css" href="https://abulka.github.io/hugo-bare1/css/style.css">
```

**Comment by epogrebnyak** [comment]
I'm struggling with the same issue - delpoying to baseURL with subdir, but canonifyurls = true breaks the menus. Discussion here: https://discourse.gohugo.io/t/figure-shortcode-cannot-find-image-in-static-folder-in-site-com-subsite-baseurl/33357/10

Wish Using hugo with Github Pages was easier. A small step in the documentatino could be mentioning how baseURL is processes - what qualifies as BaseURL, the fact that subdir may be ignored. 

There is a small discussion here: https://gohugo.io/functions/abslangurl/, but maybe users my want to hear it here: https://gohugo.io/getting-started/configuration/#all-configuration-settings

Currently we have:

```
baseURL
    Hostname (and path) to the root, e.g. https://bep.is/
```

How this can be improved? Probably by answering subdir questions:

- is `https://bep.is/subdir/` allowed as a baseURL?
- does Hugo acknowledge or ignore `subdir/` is base URL?
- what parts of functionality will not work for `subdir/`
- how to make them work and what are the tradeoffs (eg canonifyurls = true)?

My specific point of failure is to get `{{< figure src="image.png">}}` to work on `baseURL = "https://bep.is/subdir/"`, the shortcode resolves to  `https://bep.is/image.png` and not `https://bep.is/subdir/image.png`.

**Comment by amancioandre** [comment]
I've been struggling with the same issue these past couple days. Since this discussion helped half way to fix, this [topic on stackoverflow](https://stackoverflow.com/questions/55343573/generated-links-broken-in-hugo-blogdown-when-publishing-to-domain-with-subfolder) fixes the other half, however hacky this is it can help guys like me.

**Comment by epogrebnyak** [comment]
@amancioandre - feel your pain - it felt the solution is somewhere near, but still evades me and everyone is saying everythingbcan be handked easily. Glad you found both halves!

**Comment by bep** [comment]
This feels like a issue better suited for https://discourse.gohugo.io/ -- so closing this.

**Comment by simonbance** [comment]
I am having this same issue. As a workaround I set `canonifyurls = true` in the config file and manually modified the menu URLs to reflect the GitLab project path. Almost everything now works, except the "Next page" link at the bottom of the main homepage (I'm using the PaperMod theme). 
But more importantly for me, I think that `canonifyurls = true` is the reason that my custom domain name doesn't work! Although I can visit customdomain.com and be redirected to <user>github.io/<project>/, as soon as I click on any link on the site it moves it to the github version (well actually I went with GitLab pages in the end, but the same applies). 

I'm going to follow the other discussions listed from this thread and hope to find the answer.
